generator/internal/genclient/translator/protobuf/protobuf_test.go

```diff
@@ -0,0 +1,388 @@
+		t.Fatalf("Cannot find message %s in API State", ".test.Fake")
+	}
+	checkMessage(t, *message, genclient.Message{
+		Name:          "Fake",
+		Documentation: "A test message.",
+		Fields: []*genclient.Field{
+			{
+				Repeated:      true,
+				Documentation: "A repeated field tag = 1",
+				Name:          "f_double",
+				JSONName:      "fDouble",
+				ID:            ".test.Fake.f_double",
+				Typez:         genclient.DOUBLE_TYPE,
+			},
+			{
+				Repeated:      true,
+				Documentation: "A repeated field tag = 3",
+				Name:          "f_int64",
+				JSONName:      "fInt64",
+				ID:            ".test.Fake.f_int64",
+				Typez:         genclient.INT64_TYPE,
+			},
+			{
+				Repeated:      true,
+				Documentation: "A repeated field tag = 9",
+				Name:          "f_string",
+				JSONName:      "fString",
+				ID:            ".test.Fake.f_string",
+				Typez:         genclient.STRING_TYPE,
+			},
+			{
+				Repeated:      true,
+				Documentation: "A repeated field tag = 12",
+				Name:          "f_bytes",
+				JSONName:      "fBytes",
+				ID:            ".test.Fake.f_bytes",
+				Typez:         genclient.BYTES_TYPE,
+			},
+		},
+	})
+}
+
+func TestScalarOptional(t *testing.T) {
+	contents := `
+syntax = "proto3";
+package test;
+
+// A test message.
+message Fake {
+    // An optional field tag = 1
+    optional double   f_double     = 1;
+		
+	// An optional field tag = 3
+    optional int64    f_int64      = 3;
+	
+	// An optional field tag = 9
+    optional string   f_string     = 9;
+	
+	// An optional field tag = 12
+    optional bytes    f_bytes      = 12;
+}
+`
+	api := makeAPI(
+		newCodeGeneratorRequest("resources.proto", contents, t))
+
+	message, ok := api.State.MessageByID[".test.Fake"]
+	if !ok {
+		t.Fatalf("Cannot find message %s in API", "Fake")
+	}
+	checkMessage(t, *message, genclient.Message{
+		Name:          "Fake",
+		Documentation: "A test message.",
+		Fields: []*genclient.Field{
+			{
+				Optional:      true,
+				Documentation: "An optional field tag = 1",
+				Name:          "f_double",
+				JSONName:      "fDouble",
+				ID:            ".test.Fake.f_double",
+				Typez:         genclient.DOUBLE_TYPE,
+			},
+			{
+				Optional:      true,
+				Documentation: "An optional field tag = 3",
+				Name:          "f_int64",
+				JSONName:      "fInt64",
+				ID:            ".test.Fake.f_int64",
+				Typez:         genclient.INT64_TYPE,
+			},
+			{
+				Optional:      true,
+				Documentation: "An optional field tag = 9",
+				Name:          "f_string",
+				JSONName:      "fString",
+				ID:            ".test.Fake.f_string",
+				Typez:         genclient.STRING_TYPE,
+			},
+			{
+				Optional:      true,
+				Documentation: "An optional field tag = 12",
+				Name:          "f_bytes",
+				JSONName:      "fBytes",
+				ID:            ".test.Fake.f_bytes",
+				Typez:         genclient.BYTES_TYPE,
+			},
+		},
+	})
+}
+
+func newCodeGeneratorRequest(name, contents string, t *testing.T) *pluginpb.CodeGeneratorRequest {
```

Comment:
nit: generally I would expect to see t as the first parameter to the test. Also to improve test reporting should there be a failure consider marking this function as a helper function. You would call `t.Helper` at the start of the function.

Reply:
Done.